### DISCUSSION TITLE
Is there a keyboard shortcut for different highlight colors?

### DISCUSSION THREAD

**Discussion by vipatel10** [sumatrapdfreader/sumatrapdf#2594] [Ideas]
In the Kindle app, when one selects some text, a pop up shows 4 different colors to choose from for the highlight color.  This is nice because when reading, one can use different colors for different significance (blue is for vocab words, pink is something you don't understand, yellow is an important sentence/passage).  Currently, it in Sumatra one selects text and then highlights by pressing the "a" key.  If one wants to change the highlight color, one can press "A" to highlight and open the annotations editor but then one has to mouse to the color dropdown and select a color.  A bit cumbersome if one is regularly highlighting.  What do you guys think of a feature that once "a" is pressed, a second key can be pressed to change the highlight color to some custom colors that can be set by the user?

**Comment by GitHubRulesOK** [comment]
it was done in the past + other related navigating bookmark/highlighing enhancements by Nod5 via AutoHotKey in effect pressing keys would edit the smx file (no-longer used after 3.2) it can be done in a similar fashion by add a pallet to edit settings, but was naturally dropped with the change from highlight all files to only highlight pdf.

![image](https://user-images.githubusercontent.com/12754913/170144775-f0ea68e7-882c-40fe-8c39-64b6f1927faf.png)

**Reply by vipatel10** [reply]
Oh no, highlighting isn't supported in epubs anymore?  I just bought an epub and was hoping to do lots of highlighting :(  Why not supported anymore?

**Reply by vipatel10** [reply]
Oh, I see you explained here: https://forum.sumatrapdfreader.org/t/highlight-paragraphs-epub-files/4626

Are there plans to revive the epub highlight feature?  I would think that would be a popular feature no?

**Reply by kjk** [reply]
With PDF, annotations are stored inside the file.

With ePub, there is no standard for annotations that I know of, so at best they would be limited to SumatraPDF only and they would have to be stored in some external file so you would loose them when moving the file.

Not sure how useful that is.

**Reply by vipatel10** [reply]
I guess the best workaround for me would be to convert the epub to PDF and highlight the PDF so the highlights are portable.

**Reply by temberature** [reply]
calibre implements epub highlights as a file inside epub, maybe help

**Comment by runck** [comment]
Keyboard shortcut for different highlight colors is also what I need.

It think a possible and maybe the easiest way is just to add more sub-cmds of CmdCreateAnnotHighlight in the next version, whose color can be set in the advanced setting individually.
Then we can bind the shortcut to these sub-cmds if needed.

The advantages are: 1) still one key for different color; 2) insensible for other users.

Although not vey elegant, I guess more cases here would be helpful as CmdCreateAnnotHighlight_1, CmdCreateAnnotHighlight_2, etc.
https://github.com/sumatrapdfreader/sumatrapdf/blob/7c60ad96a0e96a22062d534a84bd84494b005496/src/Menu.cpp#L1746-L1748

**Comment by meteorquake** [comment]
It would be great to have more colours. Can I suggest when you press A it creates the annotation under the last colour used and if you then press A again (whilst on it) it could either cycle through the colours with each press (no popup) or alternatively just a single press when on an annotation would bring up a palette (perhaps a different key so you could have both).
I know some programs have palettes auto-come up but I do find that annoying. If you've used A to create the annotation it's not hard to press A a second time to adjust colour since your finger is already on the key.
d

**Reply by GitHubRulesOK** [reply]
hmm using keys it was done as an ahk highlight helper add on by Nod5, for at the time smx annotations (abandoned around 3.2), its possible to do similar for settings but the risk is like similar tasks needs error checking not to corrupt the settings file if other changes are afoot. 
I show a pallet above but the core was keyboard rotation of colours, my preference was since mouse is in hand its easier to click a selection.

**Comment by meteorquake** [comment]
Perhaps then it would be possible to hard-code a couple of colours to cycle through without affecting the settings as an interim :)

**Reply by GitHubRulesOK** [reply]
no my point was it required the settings to be over written as that's what's used at each A press the alternative is change after and that's what Nod5 also did by altering the smx like annots editor, but that's likely to add much bloat in a PDF each change of colure adds overheads, it needs to be before press A (pick colour > change setting > apply color)

```
before annots
>> /Root 1 0 R /Size 18 >> = 7,607 bytes
after one annot
>>/Root 1 0 R/Size 20>> replace one add two ( =3 changes = 8,858 bytes)
change colour of annot increases size of file by adding yet another description
>>/Root 1 0 R/Size 21>> replace one add one  = variable but say 9,893 - 9,900 bytes
Delete annotation
>>/Root 1 0 R/Size 21>> replace one with non visible = still there = 10,348 bytes
```

**Comment by JuninhoFreitas** [comment]
it would be like VIM commands
example:
Press 'x' and it will wait for another key, then u press 'a' and highlight with green, press 'x' + 'b' then higlight with red, etc.

how can we do it?

**Reply by GitHubRulesOK** [reply]
it works runtime press A = yellow change to `#ff0000` by overwrite second ff with 00 <kbd>Ctrl +S</kbd>ave now ALL following highlights are red, its quick and efficient and requires no programming.
![image](https://user-images.githubusercontent.com/12754913/210295250-151785d6-0795-4764-9119-b9e93920b2c0.png)

**Comment by meteorquake** [comment]
Hmm. I think Juninho's suggestion is more user-practical when you have lots of highlights as I often do, much as the other suggestion makes a useful interim if there's only one or two. So double-pressing A could just create one using the last colour used, but A followed by a different letter would choose a different colour (some useful default set). Maybe A followed by S (because it's beside the A) would bring up the pallette and you click a colour or adjust one of the letter-linked ones. It wouldn't need to be stored in settings if there's an issue there that was raised...
d

**Reply by Sieboldianus** [reply]
Would it be possible to click "a" for highlighting; and then capture any <kbd>tab</kbd> pressed afterwards to cycle through colors. The last color that was set via tab can either be set as sticky; or start with the default color on the next annotation again. Reason: I only sometimes want to highlight parts in a slightly different color, to emphasize specific parts differently.

**Comment by GitHubRulesOK** [comment]
UPDATE from tomorrow via pre-release 16374 or later https://www.sumatrapdfreader.org/prerelease

@vipatel10 @Sieboldianus @meteorquake @JuninhoFreitas @runck 

KJK has allowed shortcuts for single key set colouring (and more)
for details see https://github.com/sumatrapdfreader/sumatrapdf/issues/3607

**Reply by meteorquake** [reply]
Very happy with that solution, removing the openedit command, although taking your example pressing "A" to make yellow as before now crashes sumatra unless you create an A command too (using 16377). Let me know if that's not obviously reproducible.

**Reply by kjk** [reply]
I just fixed the crash, will be fixed in next pre-release

**Comment by meteorquake** [comment]
There is one other issue, which is that in performing a highlight the clipboard is changed to the item highlighted - is that solvable so the clipboard is untouched?

**Reply by GitHubRulesOK** [reply]
I would prefer that the "selected" text was a "feature". 
Previously the highlighted text was pasted in contents but there were complaints it needed deletion so not "auto pasted" but user can optionally paste into contents.

You can add a custom color picker as an HTA file and add commands such as add content automatically I have only updated to allow change colour by press a key before highlight as another choice 
<img width="801" height="356" alt="image" src="https://github.com/user-attachments/assets/cff36ff7-c7b7-400a-b8bc-febed085b0fc" />
<img width="806" height="277" alt="image" src="https://github.com/user-attachments/assets/6be4a518-3f2e-4a5e-8bd4-1883864a3de4" />
